From 4f1160b1b50e3c221ca7b85d35216708a0416885 Mon Sep 17 00:00:00 2001
From: Yuvasree Sivasankaran <quic_ysivasan@quicinc.com>
Date: Thu, 4 Jul 2024 16:04:16 +0530
Subject: [PATCH]QSDK: wifi: mac80211: PPE VP allocation for AP VLAN net device

In case of AP VLAN vif, PPE VP allocation is not happening.
Add change for PPE VP allocation and skip label to avoid allocation
in case of AP vif which occurs in driver.

Fixes: I5feb78a770beed5e2d031d89e832f93d036d5ff9

Patch Dependency: 910-b-wifi-mac80211-Add-framework-to-support-PPE-RFS-for-V.patch

Signed-off-by: Yuvasree Sivasankaran <quic_ysivasan@quicinc.com>
---
 net/mac80211/iface.c | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

--- a/net/mac80211/iface.c
+++ b/net/mac80211/iface.c
@@ -1388,6 +1388,8 @@ static int ieee80211_ppe_vp_802_3_redir_
 	} else if (ppe_vp_rfs && sdata->vif.type == NL80211_IFTYPE_AP_VLAN) {
 		vpai.usr_type = PPE_VP_USER_TYPE_PASSIVE;
 		vpai.core_mask = 0x7;
+	} else {
+		goto skip_ppe_vp_alloc;
 	}
 
 	sdata->vif.ppe_vp_type = ppe_vp_type;
@@ -1399,6 +1401,7 @@ static int ieee80211_ppe_vp_802_3_redir_
 	sdata_info(sdata, "Allocated PPE VP port no:%d type %d for dev:%s\n",
 		   vp, vpai.usr_type, dev->name);
 
+skip_ppe_vp_alloc:
 	return vp;
 }
 #endif
@@ -1684,7 +1687,7 @@ int ieee80211_do_open(struct wireless_de
 
 #ifdef CPTCFG_MAC80211_PPE_SUPPORT
 	/* for PPE RFS, driver handles VP allocation for non VLAN interfaces */
-	if ((ppe_vp_accel || ppe_vp_rfs) && sdata->vif.ppe_vp_num != -1) {
+	if ((ppe_vp_accel || ppe_vp_rfs) && sdata->vif.ppe_vp_num == -1) {
 		vp = ieee80211_ppe_vp_802_3_redir_vap(sdata, dev,
 						      wdev->ppe_vp_type);
 		if (vp > 0) {
